frontend/lib/wallets/WalletConnect.ts

```diff
@@ -1,32 +1,97 @@
+import { getChainIdByNetworkType, getNetworkTypeByChainId } from 'auctions-core/src/network';
+import { setSigner } from 'auctions-core/src/signer';
 import WalletConnectLogo from '~/assets/icons/wallets/walletconnect.svg';
 import AbstractWallet from '~/lib/wallets/AbstractWallet';
 
 export default class WalletConnect extends AbstractWallet {
     public static title = 'WalletConnect';
     public static icon = WalletConnectLogo;
-    public static downloadUrl = 'https://metamask.io/';
 
-    get isInterfaceReady() {
-        return true;
+    private static provider?: WalletConnectProvider;
+
+    private addresses?: string[];
+
+    public get address() {
+        if (Array.isArray(this.addresses) && this.addresses[0]) {
+            return this.addresses[0].toLowerCase();
+        }
+    }
+
+    async getProvider(): Promise<ethers.providers.JsonRpcProvider> {
+        if (!WalletConnect.provider) {
+            const rpcUrl = process.env.RPC_URL;
+            const { networks } = await setupRpcUrlAndGetNetworks(rpcUrl);
+            let rpcMap = {};
+            for (const network of networks) {
+                const chainIdDecimal = parseInt(network.chainId, 16);
+                rpcMap = {
+                    ...rpcMap,
+                    [chainIdDecimal]: network.url,
+                };
+            }
+            WalletConnect.provider = new WalletConnectProvider({ rpc: rpcMap });
+        }
+        await WalletConnect.provider.enable();
+        return new ethers.providers.Web3Provider(WalletConnect.provider);
     }
 
-    get isConnected() {
-        return true;
+    async getSigner(): Promise<ethers.providers.JsonRpcSigner> {
+        const provider = await this.getProvider();
+        await provider.send('eth_accounts', []);
+        return provider.getSigner();
     }
 
-    get address() {
-        return '';
+    public async connect(): Promise<void> {
+        const signer = await this.getSigner();
+        this.addresses = [await signer.getAddress()];
+        await this.networkChangedHandler();
+        this.setup();
     }
 
-    connect(): Promise<void> {
-        throw new Error(`${WalletConnect.title} provider is not yet implemented`);
+    public async switchNetwork(network: string): Promise<void> {
+        const provider = await this.getProvider();
+        const chainId = getChainIdByNetworkType(network);
+        await provider.send('wallet_switchEthereumChain', [{ chainId }]);
     }
 
-    switchNetwork(): Promise<void> {
-        throw new Error(`${WalletConnect.title} network switching is not yet implemented`);
+    public async networkChangedHandler() {
+        if (!WalletConnect.provider) {
+            return;
+        }
+        const signer = await this.getSigner();
+        const chainId = formatToHexWithoutPad(await signer.getChainId());
+        const networkType = getNetworkTypeByChainId(chainId);
+        if (networkType) {
+            setSigner(networkType, signer as any);
```

Comment:
Why typecasting?

Reply:
If the problem is only with the promise, you can always use `Promise.resolve()`, eg
```suggestion
            setSigner(networkType, Promise.resolve(signer));
```